src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/CachedExpressionCompiler.cs

```diff
@@ -107,17 +107,63 @@ private static class Compiler<TModel, TResult>
                     }
 
-                    var constantExpression = memberExpression.Expression as ConstantExpression;
-                    if (constantExpression != null)
+                    if (memberExpression.Expression == expression.Parameters[0])
+                    {
+                        // model => model.Member
+                        if (memberExpression.Member.DeclaringType.IsValueType &&
+                            Nullable.GetUnderlyingType(memberExpression.Type) == null)
+                        {
+                            // struct.Member
+                            return _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, _ => expression.Compile());
+                        }
+                        else if (!memberExpression.Type.IsValueType &&
+                                 Nullable.GetUnderlyingType(memberExpression.Type) != null)
+                        {
+                            // class.class, class.nullable, nullable.class, nullable.nullable
+                            return _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, _ =>
+                            {
+                                // Add null conditional
+                                var replacementMemberExpression = Expression.Lambda<Func<TModel, TResult>>(
+                                    Expression.Condition(
+                                        Expression.NotEqual(memberExpression.Expression,
+                                            Expression.Constant(null, typeof(object))),
+                                        expression,
+                                        Expression.Convert(
+                                            Expression.Constant(null, typeof(object)),
+                                            memberExpression.Type)),
+                                    expression.Parameters);
+
+                                return replacementMemberExpression.Compile();
+                            });
+                        }
+                        else
+                        {
+                            // class.struct, nullable.struct
+                            return _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, _ =>
+                            {
+                                // Add null conditional to default
+                                var replacementMemberExpression = Expression.Lambda<Func<TModel, TResult>>(
+                                    Expression.Condition(
+                                        Expression.NotEqual(memberExpression.Expression,
+                                            Expression.Constant(null, typeof(object))),
+                                        expression,
+                                        Expression.Default(memberExpression.Type)),
+                                    expression.Parameters);
```

Comment:
I'm not sure I grok this case, isn't this the case where LHS is assignable from null? I dunno why we need `Expression.Default` or if that's just ben being cute ❤️